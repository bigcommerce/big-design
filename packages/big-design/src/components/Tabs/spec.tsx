import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import React from 'react';

import 'jest-styled-components';

import { warning } from '../../utils';

import { TabItem, Tabs } from './Tabs';

const items: TabItem[] = [
  { ariaControls: 'content1', id: 'tab1', title: 'Tab 1' },
  { ariaControls: 'content2', id: 'tab2', title: 'Tab 2' },
];

jest.mock('../../utils', () => ({
  ...jest.requireActual('../../utils'),
  warning: jest.fn(),
}));

test('render Tabs', () => {
  const { container } = render(<Tabs items={items} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('render Tabs with disabled item', () => {
  const disabledItems: TabItem[] = [...items, { id: 'tab3', title: 'Tab 3', disabled: true }];
  const { container } = render(<Tabs items={disabledItems} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('Tabs has a role', () => {
  render(<Tabs items={items} />);

  expect(screen.getByRole('tablist')).toBeInTheDocument();
});

test('Tab items have a role', () => {
  render(<Tabs items={items} />);

  expect(screen.getAllByRole('tab')).toHaveLength(2);
});

test('active tab has a tabindex', () => {
  const activeTab = 'tab2';

  render(<Tabs activeTab={activeTab} items={items} />);

  const tabs = screen.getAllByRole('tab');

  tabs.forEach((tab) => {
    if (tab.id === activeTab) {
      expect(tab.tabIndex).toBe(-1);
    } else {
      expect(tab.tabIndex).toBe(0);
    }
  });
});

test('onTabClick is called', async () => {
  const onClick = jest.fn();

  render(<Tabs items={items} onTabClick={onClick} />);

  const trigger = screen.getByText('Tab 2');

  await userEvent.click(trigger);

  expect(onClick).toHaveBeenCalled();
});

test('active tab changes on tab click', async () => {
  const onClick = jest.fn();

  const { rerender } = render(<Tabs items={items} onTabClick={onClick} />);

  let trigger = screen.getByText('Tab 2');

  await userEvent.click(trigger);

  expect(onClick).toHaveBeenCalled();

  rerender(<Tabs items={items} onTabClick={onClick} />);

  trigger = screen.getByText('Tab 1');

  await userEvent.click(trigger);

  expect(onClick).toHaveBeenCalled();
});

test("clicking a disabled tab doesn't set the active tab", async () => {
  let activeTab = 'tab1';
  const disabledItems: TabItem[] = [...items, { id: 'tab3', title: 'Tab 3', disabled: true }];
  const setActiveTab = jest.fn((tabId: string) => (activeTab = tabId));

  render(<Tabs activeTab={activeTab} items={disabledItems} onTabClick={setActiveTab} />);

  const trigger = screen.getByText('Tab 3');

  await userEvent.click(trigger);

  expect(setActiveTab).not.toHaveBeenCalled();
  expect(activeTab).toBe('tab1');
});

test('does not forward styles', () => {
  const { container } = render(
    <Tabs className="test" items={items} style={{ background: 'red' }} />,
  );

  expect(container.getElementsByClassName('test')).toHaveLength(0);
  expect(container.firstChild).not.toHaveStyle('background: red');
});

test('passes the ariaControls prop to the tabs', () => {
  render(<Tabs className="test" items={items} />);

  const tabs = screen.getAllByRole('tab');

  expect(tabs[0]).toHaveAttribute('aria-controls', 'content1');
  expect(tabs[1]).toHaveAttribute('aria-controls', 'content2');
});

test('active tab has aria-selected', () => {
  render(<Tabs activeTab="tab1" items={items} />);

  const tabs = screen.getAllByRole('tab');

  expect(tabs[0]).toHaveAttribute('aria-selected', 'true');
  expect(tabs[1]).toHaveAttribute('aria-selected', 'false');
});

test('shows a warning if ariaControls is missing or fallback id does not exist', () => {
  const warningSpy = jest.fn();

  jest.mocked(warning).mockImplementation(warningSpy);

  const incompleteItems: TabItem[] = [
    { id: 'tab1', title: 'Tab 1' },
    { id: 'tab2', title: 'Tab 2' },
  ];

  render(<Tabs activeTab="tab1" items={incompleteItems} />);

  expect(warningSpy).toHaveBeenCalled();
});

test('does not warn if ariaControls is present', () => {
  const warningSpy = jest.fn();

  jest.mocked(warning).mockImplementation(warningSpy);

  render(<Tabs activeTab="tab1" items={items} />);

  expect(warningSpy).not.toHaveBeenCalled();
});

test('does not warn if fallback id is valid', () => {
  const warningSpy = jest.fn();

  jest.mocked(warning).mockImplementation(warningSpy);

  render(
    <>
      <Tabs activeTab="tab1" items={items} />
      <div id="tab1-content">Hello world</div>
    </>,
  );

  expect(warningSpy).not.toHaveBeenCalled();
});
