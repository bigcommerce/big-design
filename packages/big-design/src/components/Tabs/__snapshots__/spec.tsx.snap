// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`render Tabs 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  -webkit-transition: all 150ms ease-out;
  transition: all 150ms ease-out;
  -webkit-transition-property: background-color,border-color,box-shadow,color;
  transition-property: background-color,border-color,box-shadow,color;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: transparent;
  border-color: transparent;
  color: #3C64F4;
}

.c3:focus {
  outline: none;
}

.c3[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c3 + .bd-button {
  margin-top: 0.5rem;
}

.c3:active {
  background-color: #DBE3FE;
}

.c3:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c3:hover:not(:active) {
  background-color: #F0F3FF;
}

.c3[disabled] {
  border-color: transparent;
  color: #D9DCE9;
}

.c5 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

.c2 {
  overflow: auto;
}

.c4 {
  border: none;
  border-bottom: 0.25rem solid transparent;
  border-bottom-color: transparent;
  border-radius: 0;
  color: #3C64F4;
  pointer-events: auto;
  width: auto;
}

@media (min-width:720px) {
  .c3 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c3 {
    width: auto;
  }
}

<div
  class="c0 c1 c2"
  role="tablist"
>
  <button
    aria-controls="content1"
    aria-selected="false"
    class="c3 c4"
    id="tab1"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 1
    </span>
  </button>
  <button
    aria-controls="content2"
    aria-selected="false"
    class="c3 c4"
    id="tab2"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 2
    </span>
  </button>
</div>
`;

exports[`render Tabs with disabled item 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  -webkit-transition: all 150ms ease-out;
  transition: all 150ms ease-out;
  -webkit-transition-property: background-color,border-color,box-shadow,color;
  transition-property: background-color,border-color,box-shadow,color;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: transparent;
  border-color: transparent;
  color: #3C64F4;
}

.c3:focus {
  outline: none;
}

.c3[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c3 + .bd-button {
  margin-top: 0.5rem;
}

.c3:active {
  background-color: #DBE3FE;
}

.c3:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c3:hover:not(:active) {
  background-color: #F0F3FF;
}

.c3[disabled] {
  border-color: transparent;
  color: #D9DCE9;
}

.c5 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

.c2 {
  overflow: auto;
}

.c4 {
  border: none;
  border-bottom: 0.25rem solid transparent;
  border-bottom-color: transparent;
  border-radius: 0;
  color: #3C64F4;
  pointer-events: auto;
  width: auto;
}

@media (min-width:720px) {
  .c3 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c3 {
    width: auto;
  }
}

<div
  class="c0 c1 c2"
  role="tablist"
>
  <button
    aria-controls="content1"
    aria-selected="false"
    class="c3 c4"
    id="tab1"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 1
    </span>
  </button>
  <button
    aria-controls="content2"
    aria-selected="false"
    class="c3 c4"
    id="tab2"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 2
    </span>
  </button>
  <button
    aria-controls="tab3-content"
    aria-selected="false"
    class="c3 c4"
    disabled=""
    id="tab3"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 3
    </span>
  </button>
</div>
`;
